Add disk thresholds in the cluster state

gmarouli opened the pull request:
**Problem statement**
For a data node, we use the watermarks to determine if a node's disk usage is healthy. The watermarks can be configured in different ways and it's possible that each node has a different watermark configuration. This is not desirable, we want to use the same thresholds for all nodes and specifically the ones that the master is using.

**Proposal**
When a node is a elected as master, it will add a custom metadata to the cluster state that will describe these thresholds. For example:
```
    "health": {
      "disk": {
        "low_watermark": "85%",
        "high_watermark": "90%",
        "flood_stage_watermark": "95%",
        "frozen_flood_stage_watermark": "95%",
        "frozen_flood_stage_max_headroom": "20gb"
      }
    }
```

In this PR, we introduce the health metadata and we wire the existing disk thresholds to update the health metadata in the cluster state.

Part of https://github.com/elastic/elasticsearch/issues/84811

gmarouli committed:
Add disk thresholds in the cluster state

gmarouli commented:
@dakrone referring to https://github.com/elastic/elasticsearch/pull/87975#pullrequestreview-1022382259

I think there is a misunderstanding. The health metadata are not supposed to be node specific, we intend them to be the same for all the nodes of the cluster and determined by the master node. 

Effectively, that's what is happening right when ti comes to allocation too. All nodes have disk threshold settings (potentially different ones), but the elected master node will use their own in the allocation decider. In a similar way, we want the master node to propagate the thresholds so every node can check their disk usage and report back using the same thresholds.

Is this more clear now?

gmarouli committed:
Reformat comment

gmarouli committed:
Rename listener

gmarouli commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java:
```diff
@@ -8,26 +8,33 @@
 /**
  * A container to keep settings for disk thresholds up to date with cluster setting changes.
  */
 public class DiskThresholdSettings {
```
Keep in mind that I am considering in the next part of this PR to move this class to `org.elasticsearch.common.settings`. 

Then we can have one place with all the disk threshold settings, their validation and how to parse them. I haven't done it yet because it's going to touch a lot of files and it might get confusing. I thought it would be easier to review this, with that in mind, and do the refactoring in a single commit after we verify that the PR so far is ok.

If you prefer a different approach let me know.

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java:
```diff
@@ -278,38 +288,43 @@ private void setEnabled(boolean enabled) {
             lowWatermark,
             CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING.getKey()
         );
+        notifyListeners();
```
I don't think we need another observer mechanism in place as these methods are already listeners to settings update https://github.com/elastic/elasticsearch/pull/88175/files#diff-697233b32fbd24830c548746773a3013fc03a1cacc6ca877c8d75f0b9f68f751L121

Could the HealthMetadataService be a listener on the settings and consume the "settings changed" notifications that way?

gmarouli replied:
Fixed

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,152 @@
+                    currentHealthMetadata = taskContext.getTask().healthMetadata();
+                    taskContext.success(() -> {});
+                }
+                final var finalHealthMetadata = currentHealthMetadata;
```
is the `finalHealthMetadata` needed still?

Would checking `return currentHealthMetadata == initialHealthMetadata` below be enough?

gmarouli replied:
It's needed because of the lambda, the `currentHealthMetadata` is not final.

andreidan replied:
Ah you're right, we can't make it effectively final due to the loop :sadpanda: 
Sorry for the noise

gmarouli replied:
Indeed, I could get the last one like we did before and make it final, but I like that this is more inline with code in other similar cases.

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,152 @@
+        clusterService.submitStateUpdateTask(source, task, config, taskExecutor);
+    }
+
+    record UpdateHealthMetadataTask(HealthMetadata healthMetadata) implements ClusterStateTaskListener {
```
I'm a bit confused by this record.

It's a task but doesn't have an execute method ? Should the `createHealthMetadata` be a sort of ... `upsertHealthMedata` and be executed inside the task's execute method?

gmarouli replied:
I see. What about this:
In the health metadata class we have classes that monitor the metadata settings. The task takes this monitors and in the execute method creates the current version of the HealthMetadata. 

I implemented this idea, in my latest commit. What do you think?

andreidan replied:
Thanks for iterating on this.

Do we need this extra concept of "monitoring" though? Do we need the caching it provides for a particular use case?

Could the `HealthMetadataService` be the settings listener and trigger the necessary cluster state updates when they change? (apologies if I'm missing something)

gmarouli replied:
I do not think you are missing something. I was trying to optimize the case when we initialize the health metadata for the first time or when a master node gets elected. 

We could also parse the settings on the spot and have no cache, I was trying to avoid parsing them multiple times but on the other hand if this doesn't happen often maybe it's less important.

Do you think it would be cleaner to only track the settings?

gmarouli replied:
Settled offline, we expect that the changes will not be in high frequency so parsing the settings will not be a big burden and it remove the need of a cache.

andreidan commented on server/src/internalClusterTest/java/org/elasticsearch/health/HealthMetadataServiceIT.java:
```diff
@@ -0,0 +1,129 @@
+
+            String electedMaster = internalCluster.getMasterName();
+            {
+                HealthMetadata.DiskThresholds.DiskThreshold lowWatermark = HealthMetadata.getHealthCustomMetadata(
```
naming game :) 

This is a bit difficult to parse (at least to me) - `HealthMetadata.DiskThresholds.DiskThreshold`

Since this is in the health metadata could (should?) we come up with a new concept? Alert limits? Health limits? Disk health limits?

gmarouli replied:
I see, I have to admit that threshold sounds better than limits but I will play around with naming see if we can come up with something better.

gmarouli replied:
What about `HealthMetadata` contains `DiskMetadata` (for now it's mainly thresholds, but you never know what else we might need) and then we have the `DiskThreshold` I could also go for `DiskHealthLimits` but I am not sure if it is easier to read or not. 

About `HealthLimits` or `AlertLimits` I think it's too generic and we will have issues when we introduce other resource metadata.

Another idea I had, and I am leaning towards this tbh, is to use the `RelativeByteSizeValue` instead of `DiskThreshold`. What do you think?

I am open to other ideas :)

andreidan replied:
How about `HealthMetadata.Disk.Threshold` ? (the FQN heps removing the extra "metadata" and "disk" names)

> Another idea I had, and I am leaning towards this tbh, is to use the RelativeByteSizeValue instead of DiskThreshold. What do you think?

Are you thinking of naming here? Or the actual `RelativeByteSizeValue` entity instead of the DiskThreshold entity?

gmarouli replied:
I was thinking to use the actual `RelativeByteSizeValue` java entity since we are serializing it via string conversion. 

> How about HealthMetadata.Disk.Threshold

I do like this! I wanted to avoid generic names because it might confuse people when searching in the IDE. But since it's an inner class and it looks quite nice... I am reconsidering :) Shall we go for this?

andreidan replied:
I'm a fan :) If you like it too I say we go for this (@dakrone might have a different opinion but we can incorporate those too if they come through :) )

gmarouli replied:
I want for `HealthMetadata.Disk.Threshold` :)

gmarouli replied:
@andreidan So, I am reconsidering this choice since distributed team will be working soon(-ish) on #81406. I am proposing to already use `RelativeByteSizeValue` instead of the `Threshold`. The reason is that it has the `calculateValue` method which calculates the percentage and takes into account the `headroom`, so for the frozen tier it will take the headroom into account and for the others we will use `-1` for now until the headroom is implemented. This will make easier the transition when headroom  is implemented everywhere. Any objections?

gmarouli replied:
Proposal applied

andreidan reviewed (commented):
Thanks for working on this Mary.

I've left a few suggestions and questions

gmarouli committed:
Listen to the disk threshold settings directly

gmarouli committed:
Refactor UpdateHealthMetadataTask

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-disk-metadata-in-cluster-state

gmarouli committed:
Refactor HealthMetadata

gmarouli commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,275 @@
+            if (event.localNodeMaster() && publishedAfterElection == false) {
+                submitHealthMetadata("health-metadata-update-master-election");
+            }
+            // If the node is not the elected master anymore
```
Yes, functionally, that's exactly what I was going for. I think I cannot reuse the same code because we wait to check for the version of the cluster `event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)`. If I am not mistaken this happens after the master node is elected so we do not publish the health metadata during the start up phase of a cluster. 

Do you have any suggestions to capture this scenario?

gmarouli replied:
I will try to use `LocalNodeMasterListener`. Maybe it will produce some cleaner code.

gmarouli replied:
No, that wouldn't work because still need to check the version `event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_4_0)`.

gmarouli replied:
I am working on something, I will ping you back when it's ready :)

gmarouli committed:
Refactor triggering a HealthMetadata update

gmarouli committed:
Listen to the settings changes directly

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-disk-metadata-in-cluster-state

gmarouli committed:
Add javadoc to Threshold in HealthMetadata

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,241 @@
+        );
+    }
+
+    static class UpsertHealthMetadataTask implements ClusterStateTaskListener {
```
Shall we document this?

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,241 @@
+            this.healthMetadata = healthMetadata;
+        }
+
+        static UpsertHealthMetadataTask createInsertTask(Settings settings) {
```
Shall we document this?

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,241 @@
+            );
+        }
+
+        static UpsertHealthMetadataTask createUpdateTask(ClusterState clusterState, String setting, String value) {
```
Shall we document this?

gmarouli replied:
While documenting this, I noticed that this had a bug. The code was applying the changes on the cluster state it from the caller. Since we are working with batched updates, the last update applied would override all the previous ones. 

Fixed and now the health metadata is constructed in the execute method which makes much more sense.

andreidan replied:
++ good catch 

I've just added one more suggestion to have the `execute` method be `ClusterState -> ClusterState` (this would avoid an ignored parameter and having it @Nullable)

andreidan reviewed (approved):
LGTM, thanks for iterating on this Mary

gmarouli committed:
Test that reproduces the bug in cluster update

gmarouli committed:
Apply all the updates sequentially

gmarouli committed:
Update all watermarks per tier in the test

gmarouli committed:
Remove redundant 'public' modifier

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadata.java:
```diff
@@ -0,0 +1,329 @@
+        return PARSER.parse(parser, null);
+    }
+
+    public static HealthMetadata getHealthCustomMetadata(ClusterState clusterState) {
```
This can be nullable. Shall we mark this method with `@Nullable` ?

andreidan commented on server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java:
```diff
@@ -0,0 +1,260 @@
+            logger.error("failure during health metadata update", e);
+        }
+
+        abstract HealthMetadata execute(HealthMetadata initialHealthMetadata);
```
Should the param be marked with @Nullable?

andreidan replied:
Or maybe avoid the ignored parameter in the `Insert...` implementation by having the `ClusterState currentState` as parameter? (each implementation can then look for the `health` custom and add it/modify it)

Maybe even return `ClusterState` ?

gmarouli replied:
Ah, even better!!! Thank you for the suggestion!

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-disk-metadata-in-cluster-state

gmarouli committed:
UpsertHealthMetadataTask processes clusterState

gmarouli committed:
Use RelativeByteSizeValue in HealthMetadata

gmarouli committed:
Change the equals of the Disk thresholds

gmarouli committed:
Revert extraction of diskThresholdSettings parsing

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-disk-metadata-in-cluster-state

gmarouli committed:
Remove low watermark from HealthMetadata

gmarouli committed:
Fix HealthMetadata XContent parser

gmarouli committed:
Fix limits in random threshold generator

gmarouli commented:
_Dotting the i's and crossing the t's_
- Removed the `low watermark` from the `HealthMetadata`, we do not use it in the disk health calculation so there is no need to store it there.
- Reverted the `DiskThresholdSettingParser` in course of this PR it became unnecessary (at least for now) to decouple the settings and their parsing.

andreidan commented on server/src/main/java/org/elasticsearch/common/unit/RelativeByteSizeValue.java:
```diff
@@ -95,4 +97,17 @@ public String getStringRep() {
         }
     }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        RelativeByteSizeValue that = (RelativeByteSizeValue) o;
+        return Objects.equals(absolute, that.absolute) && Objects.equals(ratio, that.ratio);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(absolute, ratio);
+    }
```
I don't think this is needed ?

gmarouli replied:
Correct, I just saw it. Removing it.

gmarouli committed:
Remove unnecessary equals & hash code

gmarouli commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-disk-metadata-in-cluster-state
